### DISCUSSION TITLE
Custom primary key _id

### DISCUSSION THREAD

**Discussion by AlexanderTserkovniy** [strapi/strapi#6183] [Legacy Topics]
Hi there,

What I would like to clarify – is your vision as the team regarding the concept of setting the `_id` outside of any DB which Strapi supports (MongoDB, PostgreSQL, SQLite etc.), via using the headless API to push new items.

**Details**
As you might know Strapi is used in various scenarios and often it is used as additional system in the bunch of services. We at Parimatch tech heavily use it for our high load, very dynamic betting platform. It is the main CMS in our system, which is where we really appreciate your effort as we started using it from the 3.0.0-beta.13 version.

With that being said, we have our main database(s) up and running for a long time, with bunch of the services (250+) bound to its data. What is the most important that all of our collections (tables) contain the primary key set up, we reuse some of those collections in Strapi by pushing them to it. But we encountered issues with custom `_id`.

First of all, we cannot set this field during content-management plugin usage. Which is okay as we go to the model file and change it there.

But **what is the most important** is that when we want to set up any relation to any collection, which uses custom `_id` (We use MongoDB and the `_id` is not `ObjectID`) it fails with an error from Mongoose, which basically just ORM above the MongoDB. The [MongoDB itself allows](https://docs.mongodb.com/manual/core/document/#the-id-field) for a setting the custom `_id`, so it is the Mongoose issue. I wanted to create a bug report, but it appeared that straight from the fresh generation of the Strapi project and creating any collection relating to another collection with custom `_id` (set via editing the model file `*.settings.json`) issue is reproduced, so you can try it by yourself.

This last issue blocks us from using the system in all its power and we have some critical business features on hold.

So, my question is: if we contribute (following all the guidelines of course) to the `strapi-connector-mongoose` (which is where the issue comes) and it will allow the custom `_id` to be set, would you accept the Pull Request as a part of your product vision? If so, how would you want to be it done? For instance, we could create a separate section in the content-manager plugin for either string of uuid type, with a checkbox (custom `_id`) to highlight this as the preset with the exact logic behind. Or you could write here, of how do you see it done, if you accept the idea itself.

Thank you very much, appreciate quick answer.

**Comment by derrickmehaffy** [comment]
I believe they do plan to eventually support this (not just on MongoDB, but all databases as even the SQL databases support this as well).

With the core issue being all the hard coded instances of the `id` name (taking into account that Strapi is in the process of completely getting rid of `_id` in favor of `id` to make the backend less complex) as well as the ID type (objectID for mongo, and integers for SQL).

The main issue that I want to stress is if one database type (mongo) is getting this feature, then other databases should also get it **at the same time**. Previously there were disconnects in Strapi relating to features per database (GraphQL aggregations for example) and I personally feel that no database type (be that NoSQL vs SQL or even PG vs MySQL) should have an increased feature set as compared to the rest of the options (within reason of the databases naturally).

I'm sure the Strapi team would love to see a pull request (pinging @alexandrebodin to verify that) but if you do want to contribute to this feature I would ask that you also look at the bookshelf connector as well.

**Reply by derrickmehaffy** [reply]
For additional clarification, some work was started on this ages ago, but recently it was reverted and removed in: https://github.com/strapi/strapi/pull/5988

**Reply by AlexanderTserkovniy** [reply]
Thanks! We want to contribute and if Strapi official accepts it, we do it asap. Surely we will care about all DBs if necessary.

The thing with _id removing is fine, it is the same stuff with the id. The main idea is that we want it to be customizable.

**Comment by alexandrebodin** [comment]
Hi, we removed these kin dof options as the codebase is just not ready to work correctly with this kind of usecase. 

We are changing the db layer quite often for now and it would be really hard to sync on the work to do. But we would be happy to get help to support this in the future for sure :D

**Reply by derrickmehaffy** [reply]
@AlexanderTserkovniy I believe what Alex is meaning to say here is they have substantial changes planned for v4 after the stable release. Changes like what you are suggesting would likely happen there also.

**Reply by alexandrebodin** [reply]
Yes :)

**Reply by AlexanderTserkovniy** [reply]
Hi, thanks a lot.

This change is very important for us, as I already mentioned we have highly coupled on the Strapi in our codebase, but the id is generated in another DB. If you have any progress on that, or you would need any help, just ping me, we could devote some of our resources on that.

**Reply by alexandrebodin** [reply]
Sure. We can discuss this on slack if you wish to help :). 

Just out of curiosity, what motivated the use of a custom 'id' when you developed your first DB ? 
Would it be possible to handle it with a field you create in the model and using lifecycle you set it according to your need ? 


I'm thinking:

```js
beforeCreate(inputData) {
  inputData.uuid = uuid();
}
```

**Reply by AlexanderTserkovniy** [reply]
Because we already have unique primary key generated by which we can bind our other services and create indexes
